components/script/dom/bindings/structuredclone.rs

```diff
@@ -5,15 +5,129 @@
 //! This module implements structured cloning, as defined by [HTML]
 //! (https://html.spec.whatwg.org/multipage/#safe-passing-of-structured-data).
 
+use dom::bindings::conversions::root_from_handleobject;
 use dom::bindings::error::{Error, Fallible};
+use dom::bindings::js::Root;
+use dom::bindings::reflector::DomObject;
+use dom::blob::{Blob, BlobImpl};
 use dom::globalscope::GlobalScope;
-use js::jsapi::{HandleValue, MutableHandleValue};
-use js::jsapi::{JSContext, JS_ReadStructuredClone, JS_STRUCTURED_CLONE_VERSION};
-use js::jsapi::{JS_ClearPendingException, JS_WriteStructuredClone};
+use js::jsapi::{Handle, HandleObject, HandleValue, MutableHandleValue, JSAutoCompartment, JSContext};
+use js::jsapi::{JSStructuredCloneCallbacks, JSStructuredCloneReader, JSStructuredCloneWriter};
+use js::jsapi::{JS_ClearPendingException, JSObject, JS_ReadStructuredClone};
+use js::jsapi::{JS_ReadBytes, JS_WriteBytes};
+use js::jsapi::{JS_STRUCTURED_CLONE_VERSION, JS_WriteStructuredClone, JS_WriteUint32Pair};
+use js::jsapi::{MutableHandleObject, TransferableOwnership};
 use libc::size_t;
+use std::os::raw;
 use std::ptr;
 use std::slice;
 
+///TODO: Add Min and Max const to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs?
+///TODO: Determine which value Min and Max should have.
+///NOTE: Current values found at https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#323
+#[allow(dead_code)]
+#[repr(u32)]
+enum StructuredCloneTags {
+    Min = 0xFFFF8000,
+    DomBlob = 0xFFFF8001,
+    ///NOTE: To support additional types, add new tags here.
+    Max = 0xFFFFFFFF,
+}
+
+unsafe extern "C" fn read_blob(cx: *mut JSContext,
+                               r: *mut JSStructuredCloneReader,
+                               data: u32) -> *mut JSObject {
+    let length = data as usize;
+    let mut buffer = vec![0u8; length];
+    assert!(JS_ReadBytes(r, buffer.as_mut_ptr() as *mut raw::c_void, length));
+    let target_global = GlobalScope::from_context(cx);
+    /// NOTE: missing the content-type string here.
+    /// Use JS_WriteString in write_callback? Make Blob.type_string pub?
+    let blob = Blob::new(&target_global, BlobImpl::new_from_bytes(buffer), "".to_string());
+    return blob.reflector().get_jsobject().get()
+}
+
+unsafe extern "C" fn write_blob(blob: Root<Blob>,
+                                w: *mut JSStructuredCloneWriter) -> bool {
+    if let Ok(mut blob_vec) = blob.get_bytes() {
+        blob_vec.shrink_to_fit();
+        let length = blob_vec.len();
+        assert!(JS_WriteUint32Pair(w,
+                                   StructuredCloneTags::DomBlob as u32,
+                                   length as u32));
```

Comment:
I just realized now that this will truncate a blob with length greater than 2**32. We should probably pass 0 here, and call `JS_WriteUint32Pair` again with the length split up.